KSQL Testing Tool, V1 (#2802)
confluentinc/ksql: docs/testing-tool.rst

.. _ksql-testing-tool:

KSQL Testing Tool
###########################

> This describes a command line tool, the KSQL testing tool, that can be used to test a set of KSQL statements without requiring any infrastructure such as Kafka or KSQL clusters.

Reviewer: ```suggestion
Use the KSQL testing tool to test a set of KSQL statements. The KSQL testing tool 
is a command line utility that enables testing KSQL statements without requiring any infrastructure, like |ak-tm| and KSQL clusters.
```

> KSQL testing tool is a great way to design your KSQL pipeline and ensure the expected results are generated. You can colaborate on designing your KSQL statements by sharing the test files.

Reviewer: ```suggestion
The KSQL testing tool is a great way to design your KSQL pipeline and ensure that the expected results are generated. You can collaborate on designing your KSQL statements by sharing the test files.
```

> The input for the tool is a JSON file that describes a set of KSQL statements along with the input data in Kafka topic(s) and the expected output data that will also be written in Kafka topic(s) when you run the statements in a real environment.

Reviewer: ```suggestion
You provide a JSON file that describes a set of KSQL statements, along with the input data and expected output data.
```

> You can use the testing tool from terminal by callink the testing took command and passing the test file as a parameter.

Reviewer 2: ```suggestion
You can use the testing tool from terminal by calling the testing tool command and passing the test file as a parameter.
```

> You can use the testing tool from terminal by callink the testing took command and passing the test file as a parameter.

Reviewer: ```suggestion
Run the testing tool from a terminal and pass the test file as a parameter.
```

.. code:: bash

    $ ksql-testing-tool /path/to/the/test/file.json

Test File Structure
*******************

> The test file is a JSON file containing the KSQL statements, input data, desired configurations and the expected results or expected errors.

Reviewer: ```suggestion
The test file is a JSON file containing the KSQL statements, input data, desired configurations, and the expected results or expected errors.
```

> The following is a simple test file:

Reviewer 2: ```suggestion
The following is a sample test file:
```

.. code:: json
    {
      "comments": [

>         "Add a description of _what_ are of functionality this file is testing"

Reviewer: ```suggestion
        "Add a description of the functionality that this file tests"
```

            "CREATE STREAM output AS SELECT id FROM intput WHERE id < 10;"
          ],
          "inputs": [
            {"topic": "input_topic", "key": 0, "value": {"id": 8}, "timestamp": 0},
            {"topic": "input_topic", "key": 0, "value": {"id": 10}, "timestamp": 10000},
            {"topic": "input_topic", "key": 1, "value": {"id": 9}, "timestamp": 30000},
            {"topic": "input_topic", "key": 1, "value": {"id": 11}, "timestamp": 40000}
          ],
          "outputs": [
            {"topic": "OUTPUT", "key": 0, "value": {"ID": 8}, "timestamp": 0},
            {"topic": "OUTPUT", "key": 1, "value": {"ID": 9}, "timestamp": 30000}
          ]
        },
        {
          "name": "my first negative test",
          "description": "an example negative test where the statement will fail to parse",
          "statements": [
            "CREATE STREAM TEST WITH (kafka_topic='test_topic', value_format='DELIMITED');"
          ],
          "expectedException": {
            "type": "io.confluent.ksql.util.KsqlException",
            "message": "The statement does not define any columns."
          }
        }
      ]
    }

> As you can see, you can have multiple tests in one test files. In addition to name, description and statements, each test includes input topics and their data along with the expected output topics and their data.

Reviewer: ```suggestion
You can have multiple tests in one test file. In addition to name, description, and statements, each test includes input topics and their data along with the expected output topics and their data.
```

> The test file format is the same as the files KSQL code uses for query translation tests. For more details on the structure of the test file and all possible settings refer to the `README.md <https://github.com/confluentinc/ksql/tree/master/ksql-functional-tests>` in the KSQL repository.

Reviewer 2: ```suggestion
The test file format is the same as the files KSQL code uses for integration tests. For more details on the structure of the test file and all possible settings refer to the `README.md <https://github.com/confluentinc/ksql/tree/master/ksql-functional-tests>` in the KSQL repository.
```

> The test file format is the same as the files KSQL code uses for query translation tests. For more details on the structure of the test file and all possible settings refer to the `README.md <https://github.com/confluentinc/ksql/tree/master/ksql-functional-tests>` in the KSQL repository.

Reviewer: ```suggestion
The test file format is the same as the files KSQL code uses for query translation tests. For more details on the structure of the test file and all possible settings, see the `README.md <https://github.com/confluentinc/ksql/tree/master/ksql-functional-tests>`__ in the KSQL repository.
```

Running Tests
*************

> As mentioned above, you can run the tests by passing the test file to the testing tool command in terminal. Let's assume we run the above test file which is stored in test.json file in the home directory.

Reviewer: ```suggestion
Assume we run the previous test file, which is stored in test.json in the home directory.
```

> The following shows how to run the test from the terminal:

Reviewer: ```suggestion
The following command shows how to run the test from the terminal:
```

.. code:: bash

>     $ ksql-testing-tool ~/test.json

Reviewer: ```suggestion
    ksql-testing-tool ~/test.json
```

     >>> Running test: ksql-test - my first positive test
    	 >>> Test ksql-test - my first positive test passed!
     >>> Running test: ksql-test - my first negative test
    	 >>> Test ksql-test - my first negative test passed!
    All tests passed!
    $

> For each test case, the testing tool first creates and populates the input topics in it's internal simulated kafka cluster.

Reviewer 2: ```suggestion
For each test case, the testing tool first creates and populates the input topics in its internal simulated kafka cluster.
```

> For each test case, the testing tool first creates and populates the input topics in it's internal simulated kafka cluster.

Reviewer: ```suggestion
For each test case, the testing tool first creates and populates the input topics in its internal simulated Kafka cluster.
```

> It then compiles and runs the KSQL statements and finally compares the generated results with the expected ones. If the expected results are generated, the test passes, otherwise it fails.

Reviewer: ```suggestion
It compiles the KSQL statements, runs them, and compares the generated results with the expected results. If the expected results are generated, the test passes, otherwise it fails.
```

.. _ksql-testing-tool:

KSQL Testing Tool
###########################

Use the KSQL testing tool to test a set of KSQL statements. The KSQL testing tool
is a command line utility that enables testing KSQL statements without requiring any infrastructure, like |ak-tm| and KSQL clusters.
The KSQL testing tool is a great way to design your KSQL pipeline and ensure the expected results are generated.
You can collaborate on designing your KSQL statements by sharing the test files.
You provide a JSON file that describes a set of KSQL statements, along with the input data and expected output data.
Run the testing tool from a terminal and pass the test file as a parameter.

.. code:: bash

>     $ ksql-testing-tool /path/to/the/test/file.json

Reviewer: @vcrfxia That's right, now that the docs site has the nifty "copy snippet" feature.

            {"topic": "input_topic", "key": 1, "value": {"id": 9}, "timestamp": 30000},
            {"topic": "input_topic", "key": 1, "value": {"id": 11}, "timestamp": 40000}
          ],
          "outputs": [
            {"topic": "OUTPUT", "key": 0, "value": {"ID": 8}, "timestamp": 0},
            {"topic": "OUTPUT", "key": 1, "value": {"ID": 9}, "timestamp": 30000}
          ]
        },
        {
          "name": "my first negative test",
          "description": "an example negative test where the statement will fail to parse",
          "statements": [
            "CREATE STREAM TEST WITH (kafka_topic='test_topic', value_format='DELIMITED');"
          ],
          "expectedException": {
            "type": "io.confluent.ksql.util.KsqlException",
            "message": "The statement does not define any columns."
          }
        }
      ]
    }

You can have multiple tests in one test file. In addition to name, description, and statements, each test includes
input topics and their data along with the expected output topics and their data.
The test file format is the same as the files KSQL code uses for integration tests. For more details on the

> structure of the test file and all possible settings, see the `README.md <https://github.com/confluentinc/ksql/tree/master/ksql-functional-tests>` in the KSQL repository.

Reviewer 3: I think it may be better to create a new README/doc rather than linking to the README for query validation tests since the usage is different, and query validation tests support functionality that isn't supported by the testing tool (and vice versa) such as schema and topology validation, as well as the `format` and `post` attributes. We can do this in a follow-up.

Author: Good point, I'll add a new page for it. Tracking this here: #2821
